.github/workflows/msvc-full-features.yml

```diff
@@ -79,17 +92,74 @@ jobs:
       run: |
         vcpkg integrate install
 
+    - name: Download ccache
+      uses: robinraju/release-downloader@v1.4
+      with:
+        repository: 'ccache/ccache'
+        tag: 'v4.6.1'
+        fileName: 'ccache-4.6.1-windows-x86_64.zip'
+
+    - name: Install ccache
+      id: install-ccache
+      run: |
+        unzip ccache-4.6.1-windows-x86_64.zip
+        cp ccache-4.6.1-windows-x86_64/ccache.exe ccache-4.6.1-windows-x86_64/cl.exe
+        cp ccache-4.6.1-windows-x86_64/ccache.exe ccache-4.6.1-windows-x86_64/clang-cl.exe
+        mv ccache-4.6.1-windows-x86_64 ${{ env.CDDA_CCACHE_PATH }}
+
+    - name: Get ccache vars
+      id: get-vars
+      run: |
+        echo "::set-output name=datetime::$(/bin/date -u "+%Y%m%d%H%M")"
+        echo "::set-output name=ccache-path::$(echo "$APPDATA\\ccache")"
+      shell: bash
+
+    - name: ccache cache files
+      uses: actions/cache@v3
+      with:
+        path: ${{ steps.get-vars.outputs.ccache-path }}
+        # double-dash after compiler is not a typo, it is to disambiguate between g++-<date> and g++-11-<date> for restore key prefix matching
+        key: ccache-${{ github.ref_name }}-${{ runner.os }}-msvc--${{ steps.get-vars.outputs.datetime }}
+        restore-keys: |
+          ccache-master-${{ runner.os }}-msvc--
+
+    - name: Configure ccache
+      run: |
+        ${{ env.CDDA_CCACHE_PATH }}\ccache.exe -M 10G
+        ${{ env.CDDA_CCACHE_PATH }}\ccache.exe -s -v
+        ${{ env.CDDA_CCACHE_PATH }}\ccache.exe -z
+
+    - name: Symlink intermediates to C
+      run: |
+        mkdir -p "$env:TEMP\objwin"
+        New-Item -ItemType SymbolicLink -Target "$env:TEMP\objwin" objwin
```

Comment:
This is so we don't run out of disk space during the build. The uncompressed build artifacts for cdda are enormous and the ramdisk the workspace is in is not. Adding ccache to the mix put it over the top. The C drive has a lot of free space available but experiences slower i/o. Net with ccache should still be faster.

Just moving the ccache cache to the C drive might not be sufficient because embedded debug symbols into the object files (which ccache needs to work) increases the size of the outputs tremendously. The old build used a helper daemon process (the default behavior for msvc builds today) to merge debug information generated during the build into a single pdb which keeps sizes down.

Enabling NTFS compression in the ramdisk could possibly keep space usage down low enough, but incurs a cpu cost during the build due to all of the i/o going on. The machines don't have that much ram available so there's not a lot for filesystem caches to keep resident, which could lead to further slowdown during the build.

Enabling compression in ccache instead of after ccache would also slow down the build, for two reasons. One is forcing compressing each object synchronously during the build is, well, slower. Two is it prevents using hardlinks to references files inside the cache, forcing copying and decompression when writing out of cache.

The current approach puts both ccache and the intermediates into the C drive. We pay an upfront cost for decompression the whole ccache cache, but then cache hits are essentially free because it just creates a hardlink into the cache for it. Writes to cache are also basically free because ccache can hardlink the compiler output *into* cache (or, alternatively, direct the compiler to emit the output straight into cache and hardlink back out of it). There's then a large post-build compression step, but that only happens on master and after tests run, at which point we've effectively gotten all the signal we need. This means only master builds pay the bulk of the cost associated with ccache and compression and PRs get all the benefit.